components/select/Select.razor.cs

```diff
@@ -807,6 +737,89 @@ protected override void OnParametersSet()
             base.OnParametersSet();
         }
 
+        private void EvaluateDataSourceChange()
+        {
+            if (DataSource == null && _datasource == null)
+            {
+                return;
+            }
+
+            if (DataSource == null && _datasource != null)
```

Comment:
Moving from setter cause this `if` block to loose at least a part of its usefulness and also is the cause why test `Do_not_react_when_DataSource_not_changed()` is failing. In the setter it was useful when setting `DataSource` from null to a non-null value. When executed in `OnParametersSet()` it is opposite - executes when `DataSource` is set to null when it was not null previously.

Reply:
No, I dont think so. I think they should be same before and after the change. `Do_not_react_when_DataSource_not_changed()` passed because you write the `OnDataSourceChanged` attribute after `DataSource` attribute, If we move `OnDataSourceChanged` to before `DataSource`, the test will fail, like this:
![image](https://user-images.githubusercontent.com/54608128/132930184-cf6b3393-7272-49e6-b5ff-47a65eec8eb3.png)

I add `handlerCalled = false;` before `cut.SetParametersAndRender(parameters => parameters.Add(p => p.DataSource, _persons));` . Now the test will pass.

Reply:
Well...it seems you are right on both accounts. But you did clean up the `if` branch. 
There is one thing though which I do not really like in `Select` and I was wondering what is your opinion on this.
I have an issue with `OnDataSourceChange` and with `OnSelectedItemChanged`/`OnSelectedItemsChanged`, namely that they are fired during initialization of the `Select` component. I think they should not fire when the component is constructed, which not is the case now...

Reply:
Sometimes I have the same feeling. But triggering the `Change` events during initialization may be useful when the users want to inject custom logic for the first render. One example is the Table OnChange(now it is used to get remote data). Of course this smells not good, however we should provide an alternative way to inject these logics if we want to change this behavior.

Reply:
Yeah, I get that, but that custom logic should be placed in the `OnInitialized`/`OnInitializedAsync` of the component using `Select`. For me, I do have logic in `OnSelectedItemsChanged` but I have to put extra `_afterFirstRender` boolean just to avoid running the logic in there during initialization. And this is in multiple places. I just have a feeling that more often you don't have to do any logic on initialization than you do.

Reply:
I agree.